### PR TITLE
cluster-up, kind: add kind-1.23

### PR DESCRIPTION
**Pull request by zhlhahaha**
This is necessary for testing kubeVirt on Arm64 server, as, currently, Arm64 server does not support nested virtualization.

I update the default `KIND_VERSION` to 0.12.0, because kind 0.12.0 support multi-arch kindest/node:v1.23.4 image. https://github.com/kubernetes-sigs/kind/releases
~~The default `KIND_NODE_IMAGE`, quay.io/kubevirtci/kindest_node:v1.23.3, does not support Arm64~~
Update the default KIND_NODE_IMAGE to `quay.io/kubevirtci/kindest_node:v1.23.4` which support both x86_64 and Arm64
Signed-off-by: Howard Zhang <howard.zhang@arm.com>

### CODE REVIEW

**Review by jean-edouard — COMMENTED** [review]
Thank you @zhlhahaha for this, looks good to me!
/cc @dhiller 
Can we add a `check-provision` test lane for this?

File: cluster-up/cluster/kind-k8s-1.23/README.md
Comment (jean-edouard) [inline cluster-up/cluster/kind-k8s-1.23/README.md:21]: If there's only 1 node, shouldn't it also be a worker?

Comment (zhlhahaha) [reply cluster-up/cluster/kind-k8s-1.23/README.md:21]: Yes, it should be both master and worker node.

Comment (zhlhahaha) [reply cluster-up/cluster/kind-k8s-1.23/README.md:21]: But I tried to create one node cluster, and use `kubectl get nodes`. It do not show worker role.

Comment (jean-edouard) [reply cluster-up/cluster/kind-k8s-1.23/README.md:21]: I assume it can't be used for much then... We probably need something like that:
https://github.com/kubevirt/kubevirtci/blob/main/cluster-up/cluster/k8s-provider-common.sh#L153
Alternatively, I guess we could enforce a minimum of 2 nodes

Comment (zhlhahaha) [reply cluster-up/cluster/kind-k8s-1.23/README.md:21]: I am not quite sure about it. As the taint of master node is none and master node is schedulable, I can deploy kubevirt and vmi successfully on this one node cluster. Do I still need to change the label of this node?

Comment (jean-edouard) [reply cluster-up/cluster/kind-k8s-1.23/README.md:21]: Ah, got it, sounds like we're good then!

**Comment by dhiller** [comment]
@zhlhahaha just to make sure I understand the use case: can we then use kind to run the tests as we do on the existing prow-workloads cluster?

/hold

Other than that, changes look good to me.

/approve

**Comment by zhlhahaha** [comment]
> Can we then use kind to run the tests on the arm cluster as we do on the existing prow-workloads cluster? Like in: create arm presubmit jobs that test things in parallel?

For this PR, my purpose is using kind to quick deploy a development environment on Arm64, not for CI test. And I understand your idea. I think `running parallel testing in kind cluster` is a feasible way, but may need more works to do. I can open an issue for it.

> Or can we even run the arm tests on the prow-workloads cluster?

I am not sure what prow-workloads cluster is. Currently the problems for CI testing on Arm64 are:
1. Arm64 server does not nested virtualization, so the test scenario for x86_64 can not be used on Arm64.
2. Prow bootstrap image does not support Arm64, we are unable to use prow to create a kind cluster on Arm64.

**Review by dhiller — APPROVED** [review]

File: cluster-up/cluster/kind-k8s-1.23/provider.sh
Comment (enp0s3) [inline cluster-up/cluster/kind-k8s-1.23/provider.sh:18]: @zhlhahaha Following the PR description `v1.23.3` doesn't support ARM, so shouldn't we set `kindest/node:v1.23.4@sha256:0e34f0d0fd448aa2f2819cfd74e99fe5793a6e4938b328f657c8e3f81ee0dfb9` for the `KIND_NODE_IMAGE`

**Review by enp0s3 — COMMENTED** [review]
@zhlhahaha Hi! thank you for the PR. Please see my question below

File: cluster-up/cluster/kind-k8s-1.23/provider.sh
Comment (zhlhahaha) [reply cluster-up/cluster/kind-k8s-1.23/provider.sh:18]: As the kind node image can be set via environmental variable `KIND_NODE_IMAGE`, so I keep it sync with the node image in kind-1.23-vgpu. I am total fine with both image. What is your suggestion?

**Comment by oshoval** [comment]
Thanks

Maybe we should jiust bump sriov one to 1.23 and have the sriov installment there as opt-out ?
this way we will have one provider less to maintain
(unless there are more features you activated that should be opt-in)

/cc @EdDev

**Comment by oshoval** [comment]
maybe worth to simply call it `kind-1.23` instead of `kind-k8s-1.23` ?
this way it is named like the sriov one without the sriov suffix
also we know that it refer to k8s so it is redundant

**Comment by zhlhahaha** [comment]
> maybe worth to simply call it `kind-1.23` instead of `kind-k8s-1.23` ? this way it is named like the sriov one without the sriov suffix also we know that it refer to k8s so it is redundant

Updated

**Review by oshoval — APPROVED** [review]
Thanks

Just change please PR title as well

File: cluster-up/cluster/kind-1.23/provider.sh
Comment (oshoval) [inline cluster-up/cluster/kind-1.23/provider.sh:18]: Aint this version contradicts what you wrote on the PR desc?
btw if you want a multi arch image, we can mirror it to quay, now when we mirror it supports multi arch

Comment (zhlhahaha) [reply cluster-up/cluster/kind-1.23/provider.sh:18]: > Aint this version contradicts what you wrote on the PR desc? 

Do you mean that kind-1.23 contradicts with KIND_VERSION which is 0.12.0? 1.23 refer to kubernetes version rather than kind version, and here it is related to the KIND_NODE_IMAGE tag which is v1.23.3.

> btw if you want a multi arch image, we can mirror it to quay, now when we mirror it supports multi arch you need to add it here if you want https://github.com/kubevirt/project-infra/blob/main/hack/images_to_mirror.csv

Is it means that I only need to put the image name (like following line) in the csv document, then multi-arch image would automatically be pull?
```
docker.io,kindest/node:v1.23.4,quay.io/kubevirtci
```

Comment (oshoval) [reply cluster-up/cluster/kind-1.23/provider.sh:18]: > > Aint this version contradicts what you wrote on the PR desc?
> 
> Do you mean that kind-1.23 contradicts with KIND_VERSION which is 0.12.0? 1.23 refer to kubernetes version rather than kind version, and here it is related to the KIND_NODE_IMAGE tag which is v1.23.3.

"(The default KIND_NODE_IMAGE, quay.io/kubevirtci/kindest_node:v1.23.3, does not support Arm64)"
I mean that this line of the PR desc, is the actual line that you use, don't you want an image which is multi arch ?

> 
> > btw if you want a multi arch image, we can mirror it to quay, now when we mirror it supports multi arch you need to add it here if you want https://github.com/kubevirt/project-infra/blob/main/hack/images_to_mirror.csv
> 
> Is it means that I only need to put the image name (like following line) in the csv document, then multi-arch image would automatically be pull?
> 
> ```
> docker.io,kindest/node:v1.23.4,quay.io/kubevirtci
> ```

A post submit job would push it, at the first time a repo is created under quay, one of the maintainer might need to make it public, but once those two are done you will be able to use it, yes, and it will be multi arch

Comment (zhlhahaha) [reply cluster-up/cluster/kind-1.23/provider.sh:18]: I see, thanks. I will submit a PR for it.

**Comment by zhlhahaha** [comment]
As the multi-arch kind-node image have mirrored, I update the default kind node image to `quay.io/kubevirtci/kindest-node:v1.23.4`. @oshoval

**Review by oshoval — APPROVED** [review]
Thanks

Can you please update PR desc now that we do have a multi arch image?

**Comment by oshoval** [comment]
Are we good to remove the hold ?
https://github.com/kubevirt/kubevirtci/pull/784#issuecomment-1108466803
@dhiller @zhlhahaha

**Comment by oshoval** [comment]
Once https://github.com/kubevirt/kubevirtci/pull/826 is merged (if merged before this PR),
please consider to do the same change for the new provider

**Comment by dhiller** [comment]
> Once #826 is merged (if merged before this PR), please consider to do the same change for the new provider

The PR looks good to me. I think that we can add this auto detection in another PR, as this is a new provider, it will not hurt IMHO.

**Comment by xpivarc** [comment]
@dhiller `check-provision-k8s-1.23` seems to fail a lot on this PR and it is the only job holding this PR. Do you think it is worth overriding?

**Comment by zhlhahaha** [comment]
Hi @dhiller 
It seems the test `check-provision-k8s-1.23` failure is not caused by my PR.

**Comment by oshoval** [comment]
It seems that https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirtci/784/check-provision-k8s-1.23/1548619678551642112 is broken

**Comment by oshoval** [comment]
/hold

CI just test it again and again because the lanes have problems
lets either override them because they are not related to the PR, or keep hold please until they are fixed

**Pull request merged**
